### ISSUE TITLE
Missing GCS files should cause `dvc pull` to error exit

### ISSUE THREAD

**Issue by PeterJCLaw** [iterative/dvc#9867]
### Description

`dvc pull` offers an `--allow-missing` flag, which suggests that by default it will error more loudly about missing files. However if there are missing files then `dvc pull` prints a warning and exits cleanly anyway.

I've only observed this using GCS as a remote. Specifically in our case the bucket existed but the root folder had been renamed.

I couldn't reproduce this with a local file remote -- in that case I also got this output:
```console
ERROR: failed to pull data from the cloud - Checkout failed for following targets:
the-file-to-be-missing
Is your cache up to date?
<https://error.dvc.org/missing-files>
```
which doesn't appear when pulling from the broken GCS remote:

```console
2023-08-22 18:57:53,241 DEBUG: v3.15.3 (snap), CPython 3.8.10 on Linux-6.2.0-26-generic-x86_64-with-glibc2.29
2023-08-22 18:57:53,241 DEBUG: command: /snap/dvc/1411/bin/dvc pull --verbose
2023-08-22 18:57:54,734 DEBUG: Preparing to transfer data from 'gs://<bucket>/dvc/files/md5' to '/<local>/.dvc/cache/files/md5'
2023-08-22 18:57:54,734 DEBUG: Preparing to collect status from '/<local>/.dvc/cache/files/md5'
2023-08-22 18:57:54,734 DEBUG: Collecting status from '/<local>/.dvc/cache/files/md5'
2023-08-22 18:57:54,738 DEBUG: Preparing to collect status from '<bucket>/dvc/files/md5'
2023-08-22 18:57:54,738 DEBUG: Collecting status from '<bucket>/dvc/files/md5'
2023-08-22 18:57:54,738 DEBUG: Querying 1 oids via object_exists
2023-08-22 18:57:55,351 DEBUG: Estimated remote size: 4096 files
2023-08-22 18:57:55,351 DEBUG: Querying '8' oids via traverse
2023-08-22 18:57:55,656 WARNING: Some of the cache files do not exist neither locally nor on remote. Missing cache files:
md5: 4164606...62f3331a16ffc
md5: a842774...48883626b0203
md5: 64800d5...6daf115d4965e
md5: 0e4cbfc...10881ea02e05a.dir
md5: dbaf70b...1ff70e14577ba
md5: 653b717...587008bf1f150
md5: c2f9af0...69eae6133d7ad
md5: 29ba428...fa82cf0883098
Everything is up to date.
2023-08-22 18:57:55,667 DEBUG: Analytics is disabled.
```

### Environment information

**Output of `dvc doctor`:**

```console
DVC version: 3.15.3 (snap)
--------------------------
Platform: Python 3.8.10 on Linux-6.2.0-26-generic-x86_64-with-glibc2.29
Subprojects:
	dvc_data = 2.14.0
	dvc_objects = 1.0.0
	dvc_render = 0.5.3
	dvc_task = 0.3.0
	scmrepo = 1.2.1
Supports:
	azure (adlfs = 2023.8.0, knack = 0.11.0, azure-identity = 1.14.0),
	gdrive (pydrive2 = 1.17.0),
	gs (gcsfs = 2023.6.0),
	hdfs (fsspec = 2023.6.0, pyarrow = 12.0.1),
	http (aiohttp = 3.8.5, aiohttp-retry = 2.8.3),
	https (aiohttp = 3.8.5, aiohttp-retry = 2.8.3),
	oss (ossfs = 2021.8.0),
	s3 (s3fs = 2023.6.0, boto3 = 1.28.17),
	ssh (sshfs = 2023.7.0),
	webdav (webdav4 = 0.9.8),
	webdavs (webdav4 = 0.9.8),
	webhdfs (fsspec = 2023.6.0)
Config:
	System: /etc/dvc
```

**Comment by pmrowla** [comment]
There was a bug in how DVC caches remote information that is fixed in `3.15.3`, but you probably have an invalid cached result from running `dvc pull` in one of the affected releases.

When you run `dvc doctor` there should be a `Repo.site_cache_dir` listed at the end of the output. If you `rm -r` that directory and then retry the `dvc pull` from GCS it should resolve your issue.

**Comment by PeterJCLaw** [comment]
Thanks. I first observed this in our GitHub Actions CI (locally I already had a copy of the files), so I'm doubtful that there would have been anything already present in the cache directory.

Hrm, I've just spotted that our CI has DVC pinned at 3.10.1 -- might that version have been affected by the original bug?

**Comment by pmrowla** [comment]
@PeterJCLaw yes, the fix was not available until `3.12.0`. I think updating your CI to use the latest DVC release will likely resolve the issue

**Comment by PeterJCLaw** [comment]
Thanks, I'll do that.
We've also fixed the bucket so I can't easily validate this, but happy to close.
